### PR TITLE
Log 8699

### PR DESCRIPTION
**Pull request by ellynch-r7**
Add logset management support

### CODE REVIEW

File: tests/examples/misc_examples.py [hunk 1]
```diff
@@ -41,6 +41,8 @@
 
 MOCK_LOGAPI_URL = 'https://mydummyurl.com/management/logs'
 
+MOCK_LOGSETAPI_URL = 'https://mydummyurl.com/management/logsets'
```
Comment (stopal-r7) [inline tests/examples/misc_examples.py]: I think we can start removing mock urls from this file and declaring them as a constant in UT file itself. What do you think?

File: lecli/logset/commands.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    api.rename_logset(logset_id, new_name)
+
+
+@click.command(help='Available commands are:\n\t"delete_log"\n\t"add_log"')
+@click.argument('command', type=click.STRING, default=None)
+@click.argument('logset_id', type=click.STRING, default=None)
+@click.argument('log_id', type=click.STRING, default=None)
+def updatelogset(command, logset_id, log_id):
+    """Update a logset by adding or deleting a log."""
+    if command == 'add_log':
```
Comment (stopal-r7) [inline lecli/logset/commands.py:68]: these commands look very handy!

File: lecli/logset/commands.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        api.add_log(logset_id, log_id)
+    elif command == 'delete_log':
+        api.delete_log(logset_id, log_id)
+    else:
+        click.echo('Missing argument "command".')
```
Comment (stopal-r7) [inline lecli/logset/commands.py]: click library should be doing this for you. If you remove `else` part and try to run, click should get back to you with a standard error message

File: lecli/logset/commands.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+@click.argument('filename', type=click.Path(exists=True, dir_okay=False))
+def replacelogset(logset_id, filename):
+    """
+    Replace a logset of a given id with new details
+    """
+    with open(filename) as json_data:
+        try:
+            params = json.load(json_data)
+            api.replace_logset(logset_id, params)
+        except ValueError as error:
```
Comment (stopal-r7) [inline lecli/logset/commands.py]: Where can this ValueError come from?

File: lecli/logset/commands.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+@click.command()
+@click.option('-n', '--name', type=click.STRING, help="Name of new log")
+@click.option('-f', '--filename', type=click.Path(exists=True, dir_okay=False),
+              help="Full or relative path to file containing JSON log object")
+def createlogset(name=None, filename=None):
+    """
+    Create a logset with the provided name and details.
+    This method will use the JSON file first if both name and file are provided
+    """
+    if filename is not None:
```
Comment (stopal-r7) [inline lecli/logset/commands.py:20]: We are not checking the case of both name and file being provided. Should we not check that?

File: lecli/logset/api.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+
+def handle_response(response, error_message, success_code, success_message=None):
+    """Handle logset responses"""
+    if response_utils.response_error(response):
+        sys.stderr.write(error_message)
+        sys.exit(1)
+    elif response.status_code == success_code:
+        if success_message:
+            print success_message
```
Comment (stopal-r7) [inline lecli/logset/api.py]: can we use click.echo?

File: lecli/logset/api.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+from lecli import response_utils
+
+def _url():
+    """
+    Get rest query url of logset resource id.
+    """
+    return api_utils.get_management_url() + '/logsets'
+
+
+def handle_response(response, error_message, success_code, success_message=None):
```
Comment (stopal-r7) [inline lecli/logset/api.py:18]: is success_message used in anywhere?

File: lecli/logset/commands.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        with open(filename) as json_data:
+            try:
+                params = json.load(json_data)
+                api.create_logset(None, params)
+            except ValueError as error:
```
Comment (stopal-r7) [inline lecli/logset/commands.py:25]: I think it is better to catch exceptions one level down, i.e in api modules.

**Review by stopal-r7 — COMMENTED** [review]
Looks good in general, just a few changes can be done.

File: lecli/logset/commands.py (hunk 4 above)
Comment (ellynch-r7) [reply lecli/logset/commands.py]: json.load will raise a Value Error if it can't parse json_data

File: lecli/logset/commands.py (hunk 5 above)
Comment (ellynch-r7) [reply lecli/logset/commands.py:20]: We're not explicitly checking for this case. The documentation does state that we will prefer the file params if both are provided though.

If you think we should change this I can update it here & in the log commands

File: lecli/logset/api.py (hunk 6 above)
Comment (ellynch-r7) [reply lecli/logset/api.py]: Updated with sys.stdout to keep in line with the rest of the codebase

File: lecli/logset/api.py (hunk 7 above)
Comment (ellynch-r7) [reply lecli/logset/api.py:18]: It's used in the delete response

File: lecli/logset/commands.py (hunk 8 above)
Comment (ellynch-r7) [reply lecli/logset/commands.py:25]: Fixed!

File: lecli/logset/commands.py (hunk 3 above)
Comment (ellynch-r7) [reply lecli/logset/commands.py]: yep, fixed

**Review by stopal-r7 — APPROVED** [review]

**Pull request merged**